### PR TITLE
fix: reset meta to default value

### LINKED ISSUE #14198: [Bug]: Multiselect widget on reset doesn't fallback to default values.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When a multiselect widget is reset either individually or inside a form (enabling reset on submit success), the `selectedOptionValues` is resetting to `[]` (empty array) rather than to the default value.

### Steps To Reproduce

1. Drag a multiselect widget
2. Drag a Text widget and set the `Text` property to `Multiselect1.selectedOptionValues`
3. Drag a button and set the `onClick` property to reset the multiselect widget.
4. Select the value `Blue` in the multiselect.
5. Press the button to reset the widget
6. Notice the Text widget displays the  `Multiselect1.selectedOptionValues` as `[]` instead of `["GREEN", "RED"]`.

### Public Sample App

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by Rishabh-Rathod**
## Description

Earlier when we reset a widget, we used to delete all the meta properties stored in the meta reducer. 

- Expectation for reset was always to bring the widget to its initial stage. 
That is meta values should get reset to the current default value. This PR resets widget meta values to default values.

Fixes #14198

## Type of change

- Bug fix (non-breaking change which fixes an issue)

#### Cypress

- [x] [[Test Case] : Verify “Multiselect widget” reset to its default value on submit success](https://github.com/appsmithorg/TestSmith/issues/1878)
- [x] [[Test Case] : Verify “MultiSelect widget” behavior on manually deselecting the items inside widget](https://github.com/appsmithorg/TestSmith/issues/1879)
- [x] [[Test Case] : verify “Text widget” fetch data from “multiselect widget” on setting “text widget’s” property as {{ {{MultiSelect[n].selectedOptionValues}} }}](https://github.com/appsmithorg/TestSmith/issues/1880)
- [ ] https://github.com/appsmithorg/TestSmith/issues/1898

#### Jest 

- [x] metaReducer reset widget add more test case

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/reducers/entityReducers/metaReducer/index.ts [hunk 1]
```diff
@@ -93,11 +94,30 @@ export const metaReducer = createReducer(initialState, {
...
+          const defaultPropertyValue = evaluatedWidget[defaultPropertyName];
+          if (defaultPropertyValue !== undefined) {
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-expect-error
+            resetMetaObj[propertyName] = klona(defaultPropertyValue);
+          }
+        },
+      );
+    }
+
```
Comment (Rishabh-Rathod) [inline app/client/src/reducers/entityReducers/metaReducer/index.ts:118]: - Fix typescript errors and make this code properly typed.
- Also, add a test to reset.

File: app/client/src/sagas/WidgetOperationUtils.ts [hunk 2]
```diff
@@ -283,6 +283,45 @@ export function getWidgetChildren(
...
+        // TODO we need to get child widget name here somehow - denormalized data could help
+        const childWidgetName = ""; // use denormalized to get name
+        const childWidget = evaluatedDataTree[childWidgetName];
+        childrenList.push({ id: childWidgetId, widget: childWidget });
+        const grandChildren = getWidgetChildren(
+          canvasWidgets,
+          denormalizedWidgets,
+          childWidgetId,
+          evaluatedDataTree,
+        );
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/WidgetOperationUtils.ts:315]: Child reset is not yet supported. Figure out how to widget `childWidget` from evaluatedTree.

**Review by Rishabh-Rathod — COMMENTED** [review]
Next steps

File: app/client/src/sagas/WidgetOperationUtils.ts (hunk 2 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/WidgetOperationUtils.ts]: Fixed

File: app/client/src/reducers/entityReducers/metaReducer/index.ts (hunk 1 above)
Comment (Rishabh-Rathod) [reply app/client/src/reducers/entityReducers/metaReducer/index.ts]: done

File: app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx [hunk 3]
```diff
@@ -531,7 +531,7 @@ class MultiSelectWidget extends BaseWidget<
   static getDefaultPropertiesMap(): Record<string, string> {
     return {
       selectedOptions: "defaultOptionValue",
-      filterText: "",
+      // filterText: "",
```
Comment (Rishabh-Rathod) [inline app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx:534]: https://github.com/appsmithorg/appsmith/pull/10359/files#r894994806

@Tooluloope  Can you check the above comment and review this line of code change.

**Review by Rishabh-Rathod — COMMENTED** [review]
NEXT STEPS: 
- [x] Add a new cypress test for resetWidget
   - multiselect 

2. Run CI test and verify all tests working fine.

File: app/client/src/reducers/entityReducers/metaReducer/index.ts [hunk 4]
```diff
@@ -93,11 +98,27 @@ export const metaReducer = createReducer(initialState, {
...
+      Object.entries(propertyOverrideDependency).map(
+        ([propertyName, dependency]) => {
+          const defaultPropertyValue =
+            dependency.DEFAULT && evaluatedWidget[dependency.DEFAULT];
+          if (defaultPropertyValue !== undefined) {
+            resetMetaObj[propertyName] = klona(defaultPropertyValue);
+          }
+        },
+      );
+    }
```
Comment (Rishabh-Rathod) [inline app/client/src/reducers/entityReducers/metaReducer/index.ts:118]: Please add comment to explain what is happening here.

File: app/client/src/sagas/WidgetOperationUtils.ts [hunk 5]
```diff
@@ -283,6 +284,51 @@ export function getWidgetChildren(
...
+            evaluatedDataTree,
+          );
+          if (grandChildren.length) {
+            childrenList.push(...grandChildren);
+          }
+        }
+      }
+    }
+  }
+  return childrenList;
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/WidgetOperationUtils.ts:329]: Please add a comment here.

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 6]
```diff
@@ -288,7 +288,11 @@ export class AggregateHelper {
     cy.get(this.locator._widgetInDeployed(endp) + " div.rc-select-selector")
       .eq(index)
       .scrollIntoView()
-      .click();
+      .then(($element) => {
+        cy.get($element).click(+$element.width - 10, +$element.height / 2, {
+          force: true,
+        });
+      });
```
Comment (Rishabh-Rathod) [inline app/client/cypress/support/Pages/AggregateHelper.ts:295]: Please explain this maths magic with a comment.

File: app/client/src/sagas/WidgetOperationSagas.tsx [hunk 7]
```diff
@@ -701,15 +701,18 @@ const unsetPropertyPath = (obj: Record<string, unknown>, path: string) => {
...
     parentWidgetId,
+    evaluatedDataTree,
   );
-  for (const childIndex in childrenIds) {
-    const childId = childrenIds[childIndex];
-    yield put(resetWidgetMetaProperty(childId));
+
+  for (const childIndex in childrenList) {
+    const { id: childId, widget: childWidget } = childrenList[childIndex];
+    yield put(resetWidgetMetaProperty(childId, childWidget));
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/WidgetOperationSagas.tsx:715]: What if we do yield all then logic to dispatch each action in parallel?

**Review by Rishabh-Rathod — COMMENTED** [review]
Add comments

File: app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx (hunk 3 above)
Comment (Rishabh-Rathod) [reply app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx:534]: Quoting @Tooluloope "I think, it's fine to remove from the map". 
Also, as i believe that is correct way, will go ahead and remove this.

Comment (Rishabh-Rathod) [reply app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx]: done ✅

File: app/client/src/sagas/WidgetOperationSagas.tsx (hunk 7 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/WidgetOperationSagas.tsx:717]: could be tried in a separate PR

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js [hunk 8]
```diff
@@ -90,29 +94,73 @@ describe("MultiSelect Widget Functionality", function() {
...
       '[\n  {\n    "label": "Option 2",\n    "value": "2"\n  }\n]',
     );
     // Check if isDirty is set to false
-    cy.get(".t--widget-textwidget").should("contain", "false");
+    cy.get(getWidgetSelector(WIDGET.TEXT))
+      .eq(0)
+      .should("contain", "false");
+  });
+
+  it("Verify MultiSelect resets to default value", function() {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js:127]: Hey, from next time - please give test case a number - so it makes it little easier to identify a test-case in CI runs/debugging!

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js [hunk 9]
```diff
@@ -90,29 +94,73 @@ describe("MultiSelect Widget Functionality", function() {
...
+        { label: "RED", value: "RED" },
+        { label: "GREEN", value: "GREEN" },
+        { label: "YELLOW", value: "YELLOW" },
+      ]),
+    );
+    // set default value
+    cy.testJsontext(
+      "defaultvalue",
+      JSON.stringify([{ label: "RED", value: "RED" }], null, 2),
+    );
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js:142]: Can we also add few more cases -
1. Default value is more than one option value
2. Default value is empty
3. Default value outside of options value, eg: in this case ["BLUE"]

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js (hunk 8 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js:127]: I will add it to this PR itself. I unintentionally followed what was already there.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js [hunk 10]
```diff
@@ -90,29 +94,128 @@ describe("MultiSelect Widget Functionality", function() {
...
+        defaultValue: [],
+        optionsToSelect: [],
+        optionsToDeselect: [],
+      },
+      {
+        options: [
+          { label: "1", value: "1" },
+          { label: "2", value: "2" },
+          { label: "3", value: "3" },
+        ],
```
[the remark this answers is not in the document]
Comment (Rishabh-Rathod) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js]: Done

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js (hunk 9 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js]: done

**Review by Aishwarya-U-R — APPROVED** [review]
Approving cypress tests

**Review by Aishwarya-U-R — CHANGES_REQUESTED** [review]
Needs dev approval!

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js [hunk 11]
```diff
@@ -1,19 +1,26 @@
...
-const data = require("../../../../fixtures/example.json");
+import data from "../../../../fixtures/example.json";
+import {
+  PROPERTY_SELECTOR,
+  WIDGET,
+  getWidgetSelector,
+} from "../../../../locators/WidgetLocators";
+import { ObjectsRegistry } from "../../../../support/Objects/Registry";
+
+const agHelper = ObjectsRegistry.AggregateHelper,
```
Comment (rimildeyjsr) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js:15]: can we rename this to aggregateHelper? documents this variable in the latter half of the file

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 12]
```diff
@@ -228,8 +227,8 @@ export class AggregateHelper {
     cy.get(this.locator._dropDownValue(ddOption)).click();
   }
 
-  public SelectDropDown(ddOption: string, endp: string = "selectwidget") {
-    let mode = window.localStorage.getItem("inDeployedMode");
+  public SelectDropDown(ddOption: string, endp = "selectwidget") {
```
Comment (rimildeyjsr) [inline app/client/cypress/support/Pages/AggregateHelper.ts:230]: same comment here as well, if we could use better variable names here, the code would be self-documented. We can skip this if it leads to too many changes

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 13]
```diff
@@ -228,8 +227,8 @@ export class AggregateHelper {
...
+    const mode = window.localStorage.getItem("inDeployedMode");
     if (mode == "false") {
```
Comment (rimildeyjsr) [inline app/client/cypress/support/Pages/AggregateHelper.ts:232]: can we have strict equality here?

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 14]
```diff
@@ -250,17 +249,17 @@ export class AggregateHelper {
...
     index = 0,
-    endp: string = "dropdownwidget",
+    endp = "dropdownwidget",
   ) {
-    let mode = window.localStorage.getItem("inDeployedMode");
+    const mode = window.localStorage.getItem("inDeployedMode");
     //cy.log("mode frm deployed is:" + mode)
-    let modeSelector =
+    const modeSelector =
       mode == "true"
```
Comment (rimildeyjsr) [inline app/client/cypress/support/Pages/AggregateHelper.ts:259]: can we have strict equality here?

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 15]
```diff
@@ -283,12 +282,24 @@ export class AggregateHelper {
...
+    endp = "multiselectwidgetv2",
   ) {
     cy.get(this.locator._widgetInDeployed(endp) + " div.rc-select-selector")
       .eq(index)
       .scrollIntoView()
-      .click();
+      .then(($element: any) => {
+        // here, we try to click on downArrow in dropdown of multiSelect.
+        // the position is calculated from top left of the element
+        const dropdownCenterPosition = +$element.height / 2;
```
Comment (rimildeyjsr) [inline app/client/cypress/support/Pages/AggregateHelper.ts:293]: Question - what does + infront of $element mean?

File: app/client/src/reducers/entityReducers/metaReducer/index.ts [hunk 16]
```diff
@@ -93,11 +98,31 @@ export const metaReducer = createReducer(initialState, {
...
+
+      // evaluatedWidget is widget data inside dataTree, this will have latest default values of widget
+      if (evaluatedWidget) {
+        const { propertyOverrideDependency } = evaluatedWidget;
+        // propertyOverrideDependency has defaultProperty name for each meta property of widget
+        Object.entries(propertyOverrideDependency).map(
+          ([propertyName, dependency]) => {
+            const defaultPropertyValue =
+              dependency.DEFAULT && evaluatedWidget[dependency.DEFAULT];
+            if (defaultPropertyValue !== undefined) {
```
Comment (rimildeyjsr) [inline app/client/src/reducers/entityReducers/metaReducer/index.ts:118]: do we need to explicitly check for defaultPropertyValue to be undefined? else we can shorten this to just defaultPropertyValue in the if else part

File: app/client/src/sagas/WidgetOperationUtils.ts [hunk 17]
```diff
@@ -299,6 +300,54 @@ export function getWidgetChildren(
...
+  canvasWidgets: CanvasWidgetsReduxState,
+  widgetId: string,
+  evaluatedDataTree: DataTree,
+): ChildrenWidgetMap[] {
+  const childrenList: ChildrenWidgetMap[] = [];
+  const widget = _.get(canvasWidgets, widgetId);
+  // When a form widget tries to resetChildrenMetaProperties
+  // But one or more of its container like children
+  // have just been deleted, widget can be undefined
+  if (widget === undefined) {
```
Comment (rimildeyjsr) [inline app/client/src/sagas/WidgetOperationUtils.ts:318]: do we need to explicitly check for undefined here?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js (hunk 11 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/FormWidgets/MultiSelect_spec.js:15]: Rename what - the constant name? No, I prefer it be agHelper so as to not be very clumsy (as object name is also AggregateHelper) and be synchronous with all other const names.

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 12 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/AggregateHelper.ts:230]: Will change ddOption - to dropdown in next check-in

File: app/client/src/sagas/WidgetOperationUtils.ts (hunk 17 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/WidgetOperationUtils.ts:318]: The new `getWidgetChildren` function has similar logic to the older one where we were doing the same check. I maintained the same logic here.

It is possible then the widget is deleted and it won't be present in `canvasWidgets` and hence the undefined check here.
``` 
const widget = _.get(canvasWidgets, widgetId);
  // When a form widget tries to resetChildrenMetaProperties
  // But one or more of its container like children
  // have just been deleted, widget can be undefined
  if (widget === undefined) {
  ```

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 13 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/AggregateHelper.ts:232]: There is no need of strict equality here, since this variable is 'set' & 'get' only in this file. There is no user setting value to this variable directly, so do not see a need for strict equality check!

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 14 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/AggregateHelper.ts:259]: Same as above!

File: app/client/src/sagas/WidgetOperationUtils.ts (hunk 17 above)
Comment (rimildeyjsr) [reply app/client/src/sagas/WidgetOperationUtils.ts:318]: got it!

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 12 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/support/Pages/AggregateHelper.ts:230]: All the changes in AggregateHelper from `let to const` or removal of unnecessary type `:string` was autofixed by eslint in vscode.

I believe we could separately improve our test code. 
I see lot of places in test code where
- strict equality check is not used and 
- variables are not named in detail to make them self explanatory.

File: app/client/src/reducers/entityReducers/metaReducer/index.ts (hunk 16 above)
Comment (Rishabh-Rathod) [reply app/client/src/reducers/entityReducers/metaReducer/index.ts:118]: If `defaultPropertyValue` is any falsy value like `0` , `""` then `if(defaultPropertyValue)` won't help here. 

we will need to explicitly check here for the case of `undefined` only then don't update.
That put me into thought if we should allow `undefined` or not. Let me check that.

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 15 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/support/Pages/AggregateHelper.ts:293]: This was added to convert the type to number and resolve type error. I saw a same practice in other part of code and hence adopted it here also.

Screenshot 2022-06-16 at 12 12 55 PM
Screenshot 2022-06-16 at 12 13 06 PM

File: app/client/src/sagas/WidgetOperationUtils.ts (hunk 17 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/WidgetOperationUtils.ts:318]: https://github.com/appsmithorg/appsmith/pull/14468/files#r898716879

File: app/client/src/reducers/entityReducers/metaReducer/index.ts (hunk 16 above)
Comment (Rishabh-Rathod) [reply app/client/src/reducers/entityReducers/metaReducer/index.ts:118]: > That put me into thought if we should allow undefined or not. Let me check that.

metaState will already be undefined if not set any value. Hence, we could keep it like this.

File: app/client/src/sagas/WidgetOperationUtils.ts (hunk 17 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/WidgetOperationUtils.ts:318]: we could have used `if (widget)` but until we don't know all the scenarios here, we could keep it as earlier.

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 12 above)
Comment (rimildeyjsr) [reply app/client/cypress/support/Pages/AggregateHelper.ts:230]: I agree - maybe a refactor PR for this change? Can we make a backlog task in zenhub?

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 15 above)
Comment (rimildeyjsr) [reply app/client/cypress/support/Pages/AggregateHelper.ts:293]: got it

File: app/client/src/reducers/entityReducers/metaReducer/index.ts (hunk 16 above)
Comment (rimildeyjsr) [reply app/client/src/reducers/entityReducers/metaReducer/index.ts:118]: got it

**Review by rimildeyjsr — DISMISSED** [review]
LGTM

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 12 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/support/Pages/AggregateHelper.ts:230]: I think issue should be to add eslint check for test code too. That way it will make sure all the rules are followed and if we don't want any rule we could change that config. 
@Aishwarya-U-R  Do we have any plan to add eslint and prettier config in test code?

Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/AggregateHelper.ts:230]: Hey, we will need to understand that existing cypress suite was not created over-night & definitely not by one person. Lot of hands on the deck & we are trying to make it as organized as possible with new typescript cases.
Older cases/methods are too huge to change/update all of it in short span of time.

And this exercise has been on our (Automation team's) mind for sometime now - but cant guarantee any timeline for the effort as we have our hands full for new work mostly & this definitely is not the priority.

**Review by Aishwarya-U-R — COMMENTED** [review]
Cypress tests OK!

**Review by Aishwarya-U-R — APPROVED** [review]

**Review by ramsaptami — APPROVED** [review]

**Comment by ramsaptami** [comment]
- [ ] Audio widget reset does not stop playing when reset using `resetWidget` even though `playState` property changes to **NOT_STARTED**
- [ ] Triggering audio recorder reset using below code returns a success message but nothing happens to the audio recorder widget if it is already recording. But it does reset widget and it's properties when it's finished recording
`{{resetWidget("AudioRecorder1",true).then(() => showAlert('success'))}}`
- [ ] Camera widget only resets it's properties but not the UI when widget is reset. i.e., widget still shows `ok` and `discard` options after reset
- [ ] Currency widget does not reset to default values
- [ ] Tab widget does not reset default tab value correctly. Toggles between selected tab and default tab values when reset
- [ ] Similar to audio widget, video widget also only changes it's playState value but not actual UI

Widget that could not be tested:
Button, Button group, Container, Divider,  Document viewer, Filepicker, Icon, iFrame, Image (indirectly tested), List, Map chart, Menu button, Modal, Statebox, Table (partly), Text

**Comment by Rishabh-Rathod** [comment]
NOTE: With above changes behaviour of widgets is not changed when we reset it, it will work same as on release. 
Mainly what is changed here is evaluated values like `AudioRecorder1.blobURL`

> * [ ]  Audio widget reset does not stop playing when reset using `resetWidget` even though `playState` property changes to **NOT_STARTED**
> * [ ]  Similar to audio widget, video widget also only changes it's playState value but not actual UI

- Issue about audio & video widget on reset doesn't stop playing is widget's issue as it is not handling such cases of reset properly to verify we could compare behaviour on release.

> * [ ]  Camera widget only resets it's properties but not the UI when widget is reset. i.e., widget still shows `ok` and `discard` options after reset

- Seems like an issue with camera widget not handling reset.

> * [ ]  Triggering audio recorder reset using below code returns a success message but nothing happens to the audio recorder widget if it is already recording. But it does reset widget and it's properties when it's finished recording
>   `{{resetWidget("AudioRecorder1",true).then(() => showAlert('success'))}}`

-  When we start recording there is no property change in audio recorder widget until recording is stopped.
-  Audio recorder also doesn’t handle reset and also doesn’t have a `isRecording` property which could have changed on reset. 

> * [ ]  Currency widget does not reset to default values

- Checking (will update the comment)

> * [ ]  Tab widget does not reset default tab value correctly. Toggles between selected tab and default tab values when reset

- Checking  (will update the comment)


> Widget that could not be tested: Button, Button group, Container, Divider, Document viewer, Filepicker, Icon, iFrame, Image (indirectly tested), List, Map chart, Menu button, Modal, Statebox, Table (partly), Text

File: app/client/src/widgets/CurrencyInputWidget/widget/index.tsx [hunk 18]
```diff
@@ -233,7 +233,7 @@ class CurrencyInputWidget extends BaseInputWidget<
       try {
         const formattedValue = formatCurrencyNumber(
```
Comment (Rishabh-Rathod) [inline app/client/src/widgets/CurrencyInputWidget/widget/index.tsx:234]: Now when we reset a widget we reset `props.text` to `defaultText` and hence, we should be tracking changes in meta fields instead of value. 

@sbalaji1192

File: app/client/cypress/support/Pages/AggregateHelper.ts (hunk 12 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/support/Pages/AggregateHelper.ts]: replaced `ddOption` with dropdownOption here. I was not sure what `endp` stands for so i have kept it like this for now.

File: app/client/src/widgets/CurrencyInputWidget/widget/index.tsx (hunk 18 above)
Comment (Rishabh-Rathod) [reply app/client/src/widgets/CurrencyInputWidget/widget/index.tsx:234]: @sbalaji1192  could you please review this change.

**Comment by rohitagarwal88** [comment]
@yatinappsmith : Need SDET to fix test cases.
cc @somangshu @ajinkyakulkarni

**Comment by Rishabh-Rathod** [comment]
> @yatinappsmith : Need SDET help to fix test cases.
> cc @somangshu @ajinkyakulkarni 

Just to clarify we need to add tests here.

Test cases are mentioned https://github.com/appsmithorg/TestSmith/issues/1898

Please refer `AllInputWidget_spec` for an approach for how to write this test case faster. ( I can elaborate more on this to the assignee ) 

Ideal candidate to write such test cases will be one who has knowledge of most of the  selectors used for widgets and for interacting with widgets as we need to interact with 24 widgets here.

File: app/client/src/sagas/EvaluationsSaga.ts [hunk 19]
```diff
@@ -169,7 +169,7 @@ function* evaluateTreeSaga(
   if (evalMetaUpdates.length) {
     yield put(updateMetaState(evalMetaUpdates));
   }
-  log.debug({ evalMetaUpdates });
+  log.debug({ evalMetaUpdatesLength: evalMetaUpdates.length });
```
Comment (rimildeyjsr) [inline app/client/src/sagas/EvaluationsSaga.ts:172]: why are we changing this?

**Review by rimildeyjsr — APPROVED** [review]
Two questions, but otherwise LGTM

File: app/client/src/sagas/EvaluationsSaga.ts (hunk 19 above)
Comment (Rishabh-Rathod) [reply app/client/src/sagas/EvaluationsSaga.ts:172]: `evalMetaUpdates` has values typed by user and in case there is any sensitive data, we don't want to log that. 
Hence, changed this to provide enough information needed for logs.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/MultiSelect_spec copy.js [hunk 20]
```diff
@@ -1,16 +1,26 @@
+/// <reference types="Cypress" />
```
[the remark this answers is not in the document]
Comment (Rishabh-Rathod) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/MultiSelect_spec copy.js:1]: This is to load ` load type definitions that come with Cypress module`

https://docs.cypress.io/guides/tooling/typescript-support#Types-for-Custom-Commands

This is supposed to be done in `index.ts` not sure how this got added here and i don't see this locally. let me check what is wrong here.

[the remark this answers is not in the document]
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/MultiSelect_spec copy.js:1]: This line is added so VSCode enables IntelliSense for Cypress methods in this file. Sometimes in js specs it doesnt work & we add it on case basis. Should not harm anything else.

**Review by rimildeyjsr — APPROVED** [review]
